### PR TITLE
Add matrix transpose

### PR DESCRIPTION
**Pull request by geofjamg**
Signed-off-by: Geoffroy Jamgotchian <geoffroy.jamgotchian@rte-france.com>

**Please check if the PR fulfills these requirements** *(please use `'[x]'` to check the checkboxes, or submit the PR and then click the checkboxes)*
- [x] The commit message follows our guidelines
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

**Does this PR already have an issue describing the problem ?** *If so, link to this issue using `'#XXX'` and skip the rest*
No

**What kind of change does this PR introduce?** *(Bug fix, feature, docs update, ...)*
Feature

**What is the current behavior?** *(You can also link to an open issue here)*
There is no way to transpose a matrix.

**What is the new behavior (if this is a feature change)?**
We can transpose a matrix either for the dense or sparse implementation.

**Does this PR introduce a breaking change or deprecate an API?** *If yes, check the following:*
- [ ] The *Breaking Change* or *Deprecated* label has been added
- [ ] The migration guide has been updated in the github wiki *(What changes might users need to make in their application due to this PR?)*

**Other information**:

(if any of the questions/checkboxes don't apply, please delete them entirely)

### CODE REVIEW

File: math/src/main/java/com/powsybl/math/matrix/DenseMatrix.java [hunk 1]
```diff
@@ -260,6 +260,18 @@ protected int getEstimatedNonZeroValueCount() {
...
+    @Override
+    public DenseMatrix transpose() {
+        double[] transposed = new double[columnCount * rowCount];
+        for (int i = 0; i < rowCount; i++) {
+            for (int j = 0; j < columnCount; j++) {
+                double value = buffer.getDouble(j * Double.BYTES * rowCount + i * Double.BYTES);
+                transposed[i * columnCount + j] = value;
+            }
+        }
+        return new DenseMatrix(columnCount, rowCount, transposed);
```
Comment (flo-dup) [inline math/src/main/java/com/powsybl/math/matrix/DenseMatrix.java:272]: ouch, not an implementation meant to be very effective, but we have here a `columnCount * rowCount` temporary array. If we use the `set(int, int, double)`, we avoid that, at the cost of the `checkBounds` for each though.

File: math/src/main/java/com/powsybl/math/matrix/SparseMatrix.java [hunk 2]
```diff
@@ -385,6 +385,15 @@ protected int getEstimatedNonZeroValueCount() {
         return values.size();
     }
 
+    private native SparseMatrix transpose(int m, int n, int[] ap, int[] ai, double[] ax);
+
+    @Override
+    public SparseMatrix transpose() {
+        SparseMatrix transposed = transpose(rowCount, columnCount, columnStart, rowIndices.getData(), values.getData());
+        transposed.setRgrowthThreshold(rgrowthThreshold);
```
Comment (flo-dup) [inline math/src/main/java/com/powsybl/math/matrix/SparseMatrix.java:393]: Just noticed that we're setting the same r growth threshold to all matrix, which questions me about it being a parameter of SparseMatrix.

File: math/src/main/java/com/powsybl/math/matrix/DenseMatrix.java (hunk 1 above)
Comment (geofjamg) [reply math/src/main/java/com/powsybl/math/matrix/DenseMatrix.java]: Right, fixed by directly filling a bytebuffer, we can avoid unnecessary bounds check.

**Review by flo-dup — APPROVED** [review]

File: math/src/main/java/com/powsybl/math/matrix/SparseMatrix.java (hunk 2 above)
Comment (geofjamg) [reply math/src/main/java/com/powsybl/math/matrix/SparseMatrix.java:393]: We just consider that a derivated matrix (transposed, multiplied) has by default the same parameter value than the initial one which seems to be fine to me. The rgrowthThreshold can be defined in the SparseMatrixFactory, so all matrices created from the same factory have by default the same value.

**Pull request merged**